nat/src/stateful/mod.rs

```diff
@@ -370,32 +286,58 @@ impl StatefulNat {
         // - tuple r.init = (src: f.nated.dst, dst: f.nated.src)
         // - mapping r.nated = (src: f.init.dst, dst: f.init.src)
 
-        let (reverse_src_addr, reverse_src_port) =
+        let (reverse_src_addr, allocated_src_port_to_use) =
             match alloc.dst.as_ref().map(|a| (a.ip(), a.port())) {
-                Some((ip, port)) => (ip, Some(port.as_u16())),
+                Some((ip, port)) => (ip.to_ip_addr(), Some(port)),
                 // No destination NAT for forward session:
                 // f.init:(src: a, dst: b) -> f.nated:(src: A, dst: b)
                 //
                 // Reverse session will be:
                 // r.init:(src: b, dst: A) -> r.nated:(src: b, dst: a)
                 //
                 // Use destination IP and port from forward tuple.
-                None => (tuple.dst_ip, tuple.dst_port),
+                None => (*flow_key.data().dst_ip(), None),
             };
-        let (reverse_dst_addr, reverse_dst_port) =
+        let (reverse_dst_addr, allocated_dst_port_to_use) =
             match alloc.src.as_ref().map(|a| (a.ip(), a.port())) {
-                Some((ip, port)) => (ip, Some(port.as_u16())),
-                None => (tuple.src_ip, tuple.src_port),
+                Some((ip, port)) => (ip.to_ip_addr(), Some(port)),
+                None => (*flow_key.data().src_ip(), None),
             };
 
-        let reverse_tuple = NatTuple::new(
+        let reverse_proto_key = match flow_key.data().proto_key_info() {
```

Comment:
There are already reverse methods for flow_key, (e.g., FlowKey::reverse) and corresponding methods for the proto key.  Is there a reason these are not acceptable?